### DISCUSSION TITLE
Demo✨: Using mswjs to create serverside & client side mocks + work on your app offline

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [remix-run/remix#4982] [Show and tell / tips]
Created an example repo of running [MSW](https://mswjs.io/) on the server. 
If you're unfamiliar with MSW its a request interception library that makes mocking responses on the server or browser easier. 

### Demo
https://user-images.githubusercontent.com/6743796/210154192-580e9f19-7be8-4f72-82a5-5152a7a6e313.mp4

demo video showcases:
- making an API call to example.com, and MSW intercepting the request and returning a response.
- running playwright test suite, it runs my app as a user would, fetches data, but recieved the mocked responses from MSW

### Highlights
- work entirely offline on your app with the help of MSW (as long as you setup the handlers before going offline)
- run your E2E/Integration test suite thats powered by MSW a lot faster by using MSW
- ability to share your handler code on the server or browser

## References
- Example code & repo: https://github.com/cliffordfajardo/remix-msw-node-with-playwright 
- MSW repo: https://github.com/mswjs/msw#awards--mentions

**Comment by runofthemill** [comment]
This is great! A note for anyone who's looking to try it out that the demo uses `msw@next` - it took me a little while to realize it wouldn't work with the current stable release (`0.49.x` at the moment)

**Comment by maxfi** [comment]
Just wanted to say thanks for this @cliffordfajardo! 🙏 This was a great help to get `msw` setup in my remix app. 🎉

This should probably be added to https://github.com/remix-run/examples/tree/main/msw. 🚀

I believe that starting the `msw` server in `app/entry.server.tsx` is a better integration than the official example where the javascript (not typescript) mock server/handlers are required as part of the package.json dev command: `binode --require ./mocks -- @remix-run/dev:remix dev`.

Also, from a quick test, it appears that the `msw` related code isn't included in the prod bundle 🎉 when initialised like this:

```tsx
// app/entry.server.tsx
import { startMockServer } from './msw-server'
import { peopleHandlers} from './msw-handlers'

if (process.env.NODE_ENV === "development" || process.env.NODE_ENV === "test") { 
  startMockServer([...peopleHandlers]);
}
```
